### PR TITLE
Seperable Convolution with Asymmetric Kernels + Fast Gauss

### PR DESCRIPTION
**Pull request by maarzt**
Separable asymmetric kernels are used in image processing.
One example: The derivative of a gauss kernel is a separable, asymmetric kernel, and is used to calculate exact derivatives of an gaussian blurred image.

This PR deprecates the current separable symmetric convolution algorithm. And adds generalized algorithm for separable convolution. The new algorithm has been benchmarked against the current implementation an proved to be just as fast.

Gauss3 is changed to use the new implementation.

### CODE REVIEW

**Comment by maarzt** [comment]
Do we want to move the classes for separable (symmetric) convolution to `net.imglib2.algorithm.convolution`?

**Comment by maarzt** [comment]
I think this PR is ready to be merged and I need it to implement features for pixel classification.

File: src/main/java/net/imglib2/algorithm/convolution/fast_gauss/FastGauss.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+
+/**
+ * Faster alternative to {@link net.imglib2.algorithm.gauss3.Gauss3}.
+ * It's especially faster if sigma is larger than 3.
+ * <p>
+ * The result is a good approximation to "real" gauss. But the
+ * same holds for {@link net.imglib2.algorithm.gauss3.Gauss3}.
+ * Which one is better is not known so far.
```
Comment (tpietzsch) [inline src/main/java/net/imglib2/algorithm/convolution/fast_gauss/FastGauss.java]: This comment is not very helpful IMO. Would be better to state what the difference in approximation is.
I.e. `Gauss3` does a (non-approximated) convolution with a Gaussian kernel truncated at 3*sigma+1.
What is the approximation here?

Also, you should add a reference to the paper that is implemented here.

File: src/main/java/net/imglib2/algorithm/gauss3/Gauss3.java [hunk 2]
```diff
@@ -163,7 +165,7 @@
 	{
 		final double[][] halfkernels = halfkernels( sigma );
 		final ExecutorService service = Executors.newFixedThreadPool( numThreads );
-		SeparableSymmetricConvolution.convolve( halfkernels, source, target, service );
+		SeparableKernelConvolution.convolve( Kernel1D.symmetric( halfkernels ), source, target );
```
Comment (tpietzsch) [inline src/main/java/net/imglib2/algorithm/gauss3/Gauss3.java]: This does not use the provided `ExecutorService`

File: src/main/java/net/imglib2/algorithm/gauss3/Gauss3.java [hunk 3]
```diff
@@ -200,7 +202,7 @@
 	public static < S extends NumericType< S >, T extends NumericType< T > > void gauss( final double[] sigma, final RandomAccessible< S > source, final RandomAccessibleInterval< T > target, final ExecutorService service ) throws IncompatibleTypeException
 	{
 		final double[][] halfkernels = halfkernels( sigma );
-		SeparableSymmetricConvolution.convolve( halfkernels, source, target, service );
+		SeparableKernelConvolution.convolve( Kernel1D.symmetric( halfkernels ), source, target );
```
Comment (tpietzsch) [inline src/main/java/net/imglib2/algorithm/gauss3/Gauss3.java]: This does not use the provided `ExecutorService`

File: src/main/java/net/imglib2/algorithm/convolution/kernel/SeparableKernelConvolution.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+ * Convolution with a customizable separable kernel.
+ *
+ * @author Matthias Arzt
+ */
+public class SeparableKernelConvolution
+{
+	/**
+	 * Return an object, that performs the separable convolution with the give kernel.
+	 * It additional allows to set the {@link java.util.concurrent.ExecutorService} to use
+	 * or to query for the required input image size, ...
```
Comment (tpietzsch) [inline src/main/java/net/imglib2/algorithm/convolution/kernel/SeparableKernelConvolution.java]: Typos: "give" --> "given"
"additional" --> "additionally"

**Comment by maarzt** [comment]
@tpietzsch I made all the changes you required. And finally I cited the paper describing the fast gauss algorithm that is implemented.

**Comment by tpietzsch** [comment]
@maarzt Did you push your changes? `Gauss3` still doesn't use user-provided `ExecutorService`

**Comment by maarzt** [comment]
@tpietzsch I did the changes you asked for:
1. Intermediate results for convolution of all RealTypes except DoubleType is now FloatType.
2. I use assertImageEquals from new imglib2-5.6.0

**Comment by tpietzsch** [comment]
Great, thanks!

**Pull request merged**